File: articles/biology-scientific-understanding-living-order/python/feedback_dynamics.py
"""
Feedback dynamics workflow.

Run:
    python python/feedback_dynamics.py
"""

from __future__ import annotations

from pathlib import Path

import pandas as pd

from living_order_core import feedback_response


ARTICLE_DIR = Path(__file__).resolve().parents[1]
FEEDBACK_PATH = ARTICLE_DIR / "data" / "feedback_scenarios.csv"


def main() -> None:
    """Calculate corrective response for feedback scenarios."""

    feedback = pd.read_csv(FEEDBACK_PATH)

    feedback["deviation"] = feedback["state"] - feedback["setpoint"]
    feedback["corrective_response"] = [
        feedback_response(state, setpoint, gain)
        for state, setpoint, gain in zip(
            feedback["state"],
            feedback["setpoint"],
            feedback["feedback_gain"],
        )
    ]

    feedback["absolute_response"] = feedback["corrective_response"].abs()

    print(feedback.round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/biology-scientific-understanding-living-order/r/feedback_dynamics.R
# Feedback dynamics in R.

feedback_path <- file.path("data", "feedback_scenarios.csv")

if (!file.exists(feedback_path)) {
  feedback_path <- file.path("..", "data", "feedback_scenarios.csv")
}

feedback <- read.csv(feedback_path)

feedback$deviation <- feedback$state - feedback$setpoint
feedback$corrective_response <- feedback$feedback_gain * (feedback$setpoint - feedback$state)
feedback$absolute_response <- abs(feedback$corrective_response)

print(round(feedback, 5))
